flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/table/PubSubTableSourceFactory.java

```diff
@@ -0,0 +1,78 @@
+package org.apache.flink.streaming.connectors.gcp.pubsub.table;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.connector.format.DecodingFormat;
+import org.apache.flink.table.connector.source.DynamicTableSource;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.factories.DeserializationFormatFactory;
+import org.apache.flink.table.factories.DynamicTableSourceFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+import org.apache.flink.table.types.DataType;
+
+import java.util.HashSet;
+import java.util.Set;
+
+@Internal
+public class PubSubTableSourceFactory implements DynamicTableSourceFactory {
+
+    // define all options statically
+    public static final ConfigOption<String> PROJECT_NAME =
+            ConfigOptions.key("projectName").stringType().noDefaultValue();
+
+    public static final ConfigOption<String> TOPIC =
+            ConfigOptions.key("topic").stringType().noDefaultValue();
+
+    public static final ConfigOption<String> FORMAT =
+            ConfigOptions.key("format").stringType().noDefaultValue();
+
+    @Override
+    public String factoryIdentifier() {
+        return "pubsub";
+    }
+
+    @Override
+    public Set<ConfigOption<?>> requiredOptions() {
+        final Set<ConfigOption<?>> options = new HashSet<>();
+        options.add(PROJECT_NAME);
+        options.add(TOPIC);
+        options.add(FORMAT);
+        return options;
+    }
+
+    @Override
+    public Set<ConfigOption<?>> optionalOptions() {
+        return new HashSet<>();
+    }
+
+    @Override
+    public DynamicTableSource createDynamicTableSource(Context context) {
+        // either implement your custom validation logic here ...
+        // or use the provided helper utility
+        final FactoryUtil.TableFactoryHelper helper =
+                FactoryUtil.createTableFactoryHelper(this, context);
+
+        // validate all options
+        helper.validateExcept("json.");
```

Comment:
There are no `json.*` options? What is this doing?

Reply:
When there is json config in the create table sql, the validation here failed.
These are the json config I see:
```
 -- 'json.map-null-key.mode' = 'FAIL', -- Optional flag to control the handling mode when serializing null key for map data, FAIL by default. Option DROP will drop null key entries for map data. Option LITERAL will use 'map-null-key.literal' as key literal.
  -- 'json.encode.decimal-as-plain-number' = 'false', -- Optional flag to specify whether to encode all decimals as plain numbers instead of possible scientific notations, false by default.
  -- 'json.map-null-key.literal' = 'null', -- Optional flag to specify string literal for null keys when 'map-null-key.mode' is LITERAL, "null" by default.
  -- 'json.fail-on-missing-field' = 'false', -- Optional flag to specify whether to fail if a field is missing or not, false by default.
  -- 'json.ignore-parse-errors' = 'false', -- Optional flag to skip fields and rows with parse errors instead of failing; fields are set to null in case of errors, false by default.
  -- 'json.timestamp-format.standard' = 'SQL', -- Optional flag to specify timestamp format, SQL by default. Option ISO-8601 will parse input timestamp in "yyyy-MM-ddTHH:mm:ss.s{precision}" format and output timestamp in the same format. Option SQL will parse input timestamp in "yyyy-MM-dd HH:mm:ss.s{precision}" format and output timestamp in the same format.
  ```
Any other idea how to address it?

Reply:
Can you post the full DDL you're using that causes errors? When you specify `'format' = 'json'`, the respective format options should be considered automatically.

Reply:
The DDL:
```
CREATE TABLE `vvp`.`default`.`my_table` (
  `user_id` VARCHAR(2147483647),
  `country_code` VARCHAR(2147483647),
  `event` VARCHAR(2147483647),
  `os` VARCHAR(2147483647)
)
COMMENT ''
WITH (
  'connector' = 'pubsub',
  'format' = 'json',
  'json.map-null-key.literal' = 'null',
  'json.map-null-key.mode' = 'LITERAL',
  'projectName' = 'myProj,
  'topic' = 'mySubscription'
);
```

The error I get (if I don't exclude the 'json' from the validation):

```
Invalid connector configuration
The source connector for table 'vvp.default.my_table' could not be created. Configured properties are: ['connector'='pubsub', 'format'='json', 'json.map-null-key.literal'='null', 'json.map-null-key.mode'='LITERAL', 'projectName'='myProj', 'topic'='mySubscription'].
Cause: Unsupported options found for 'pubsub'.

Unsupported options:

json.map-null-key.literal
json.map-null-key.mode
```

Reply:
Thanks for sharing that information. We cannot just exclude `json.*` because if you use another format, the format options would also have a different prefix.

The reason this fails is because we're currently calling things in the wrong order. You can move the call to `helper.discoverDecodingFormat` to before (what will now be) `helper.validate()`. This will register the options of the discovered format internally so they're ignored when validating the connector options.

Reply:
Got it, Thanks for the explanation!
I fixed it [in this revision](https://github.com/apache/flink/pull/16598/commits/e6b0b05fadd9d6724a424c313a7b9617a854955f)